wallet/src/account/utxo_selector/mod.rs

```diff
@@ -0,0 +1,695 @@
+        else if curr_selection_weight > max_weight {
+            // at least one selection attempt exceeded the max weight
+            max_tx_weight_exceeded = true;
+            true
+        }
+        // Selected value is within range
+        else if curr_value >= selection_target {
+            // This is the excess value which is added to the waste for the below comparison
+            let curr_waste = (curr_value - selection_target)
+                .expect("curr_value is larger than target")
+                .into_signed()
+                .and_then(|difference| curr_waste + difference)
+                .ok_or(UtxoSelectorError::AmountArithmeticError)?;
+            // Adding another UTXO after this check could bring the waste down if the long term fee is higher than the current fee.
+            // However we are not going to explore that because this optimization for the waste is only done when we have hit our target
+            // value. Adding any more UTXOs will be just burning the UTXO; it will go entirely to fees. Thus we aren't going to
+            // explore any more UTXOs to avoid burning money like that.
+            if curr_waste <= best_waste {
+                best_selection = curr_selection.clone();
+                best_waste = curr_waste;
+            }
+            true
+        } else {
+            false
+        };
+
+        // Backtracking, moving backwards
+        if backtrack {
+            let curr_last_selected_utxo_index = match curr_selection.pop() {
+                // We have walked back to the first utxo and no branch is untraversed. All solutions searched
+                None => break,
+                Some(idx) => idx,
+            };
+
+            // Add omitted UTXOs back to lookahead before traversing the omission branch of last included UTXO.
+            let new_available_value = utxo_pool[curr_last_selected_utxo_index + 1..utxo_pool_index]
+                .iter()
+                .map(|utxo| utxo.get_effective_value(pay_fees))
+                .sum::<Option<Amount>>()
+                .expect("sum of all UTXOs has already been checked to not overflow");
+
+            curr_available_value =
+                (curr_available_value + new_available_value).expect("no overflow");
+            // Output was included on previous iterations, try excluding now.
+            utxo_pool_index = curr_last_selected_utxo_index;
+
+            let utxo = &utxo_pool[utxo_pool_index];
+            curr_value = (curr_value - utxo.get_effective_value(pay_fees)).expect("no underflow");
+            curr_waste = utxo
+                .fee
+                .into_signed()
+                .and_then(|fee| {
+                    utxo.long_term_fee.into_signed().and_then(|long_term_fee| fee - long_term_fee)
+                })
+                .and_then(|fee_difference| curr_waste - fee_difference)
+                .ok_or(UtxoSelectorError::AmountArithmeticError)?;
+            curr_selection_weight -= utxo.weight;
```

Comment:
I would say it would probably be nicer to abstract these into a separate struct that represents a partial solution. Any time a utxo is added to or removed from the partial solution, it updates all the metrics like the value, waste, weight etc. Maybe it would even be possible to reuse it in multiple solvers but not sure about that. It seems the main algorithm is made less clear by being interleaved with substantial amount of bookkeeping.